tests/Agency/StorePerformanceTest.cpp

```diff
@@ -0,0 +1,423 @@
+    try {
+      auto r{write(json)};
+      auto applied_all = std::all_of(r.begin(), r.end(), [](auto const& result) {
+        return result == consensus::apply_ret_t::APPLIED;
+      });
+      if (!applied_all) {
+        throw std::runtime_error("This didn't work: " + json);
+      }
+    } catch (std::exception& ex) {
+      throw std::runtime_error(std::string(ex.what()) + " processing " + json);
+    }
+  }
+
+  void assertEqual(std::shared_ptr<velocypack::Builder> result,
+                   std::string const& expected_result) const {
+    try {
+      auto expected{VPackParser::fromJson(expected_result)};
+      if (!velocypack::NormalizedCompare::equals(result->slice(), expected->slice())) {
+        throw std::runtime_error(
+            result->toJson() + " should have been equal to " + expected->toJson());
+      }
+    } catch (std::exception& ex) {
+      throw std::runtime_error(std::string(ex.what()) + " comparing to " + expected_result);
+    }
+  }
+
+  template <typename TSource>
+  static std::string to_json_object(
+      TSource const& src,
+      std::function<std::string(typename TSource::const_reference)> extractName,
+      std::function<std::string(typename TSource::const_reference)> extractValue) {
+    bool is_first{true};
+    return "{" +
+           std::accumulate(src.begin(), src.end(), std::string(),
+                           [&is_first, extractName,
+                            extractValue](std::string partial, auto const& element) {
+                             if (is_first) {
+                               is_first = false;
+                             } else {
+                               partial += ",";
+                             }
+                             partial += "\"" + extractName(element) +
+                                        "\": " + extractValue(element);
+                             return partial;
+                           }) +
+           "}";
+  }
+
+  template <typename TSource>
+  static std::string to_json_object(TSource const& src) {
+    return to_json_object(
+        src, [](auto const& element) { return element.first; },
+        [](auto const& element) { return element.second; });
+  }
+};
+
+// hammer a single key deep in the hierarchy 100000 times or so.
+//  We test long paths.
+//  Performance relative to depth.
+
+TEST_F(StorePerformanceTest, single_deep_key_writes) {
+  std::map<std::string, std::string> obj{};
+  for (int i{}; i < hammer_times; ++i) {
+    obj["a/b/c/d/e/f/g/h"] = std::to_string(i);
+    write(std::vector<std::vector<std::string>>{{to_json_object(obj)}});
+  }
+}
+
+TEST_F(StorePerformanceTest, single_deep_key_writes_and_reads) {
+  std::map<std::string, std::string> obj{};
+  auto const key {"a/b/c/d/e/f/g/h"};
+  auto const json_read_key {std::string("[[\"") + key + "\"]]"};
+  for (int i{}; i < hammer_times; ++i) {
+    obj[key] = std::to_string(i);
+    write(std::vector<std::vector<std::string>>{{to_json_object(obj)}});
```

Comment:
Better take the construction of the JSON string and the vectors out of the loop, we only want to measure the Store operations.

Reply:
Agreed! I can spare a bit of memory for it.